### ISSUE TITLE
Overlay default operation on array items is append

### ISSUE THREAD

**Issue by joaopapereira** [k14s/ytt#269]
## Story

**As a** Configuration Consumer
**I want** `ytt` when I specify an array item in an overlay, that unless I indicate otherwise, `ytt` should assume I mean to appending that item to the existing array
**so that** when I perform this common edit, I don't have to explicitly instruct `ytt` to do so.

## Context

We've been talking about possibilities for default operations for arrays. That _general_ discussion is captured in #254.

A core goal of the imminent Schema feature is a significantly improved UX for Configuration Consumers. Users broadly report frustrating starts with `ytt` largely due to having to learn the overlay feature, immediately.
Ideally, users need only learn of the `@data/values` annotation for most configuration cases. Providing a default operation for array items gets us there.

---

### 🐰  Implied append 🎩

**Given** an overlay (aka "right") that contains an array item
**And** that array item has no `@overlay/...` annotation
**When** that overlay executes
**Then** the result will include that array item appended as if it had been annotation with a `@overlay/append`

#### Example 1: Non-empty "left" and no explicit operations in "right"

`base.yml`
```yaml
---
list:
- default value 1
- default value 2
```

`overlay.yml`
```yaml
#@ load("@ytt:overlay", "overlay")
#@overlay/match by=overlay.all
---
list:
- my value A
- my value B
```

```
$ ytt -f base.yml -f overlay.yml
---
list:
- default value 1
- default value 2
- my value A
- my value B
```

#### Example 1b: Empty array in "left"
`base.yml`
```yaml
list: []
```

`overlay.yml`
```yaml
#@ load("@ytt:overlay", "overlay")
#@overlay/match by=overlay.all
---
list:
- value A
- value B
```

```
$ ytt -f base.yml -f overlay.yml
---
list:
- value A
- value B
```

#### Example 1c: Empty array in "right"
`base.yml`
```yaml
list:
- default value 1
- default value 2
```

`overlay.yml`
```yaml
#@ load("@ytt:overlay", "overlay")
#@overlay/match by=overlay.all
---
list: []    # <-- no array items, no operations
```

```
$ ytt -f base.yml -f overlay.yml
---
list:
- default value 1
- default value 2
```

### 🗣  Explicit operation👂

**Given** an overlay (aka "right") that contains an array item
**And** it has one or more `@overlay/...` annotations on it
**When** that overlay executes
**Then** the result will be to overlay the "right" on the "left" literally as written (i.e. no implied behaviors)

#### Example 2a: Implied replace

`base.yml`
```yaml
---
list:
- default value 1
- default value 2
```

`overlay.yml`
```yaml
#@ load("@ytt:overlay", "overlay")
#@overlay/match by=overlay.all
---
list:
- my value A
#@overlay/match by=overlay.index(1)
- my value B
```

```
$ ytt -f base.yml -f overlay.yml
---
list:
- default value 1
- my value B
- my value A
```

### 👓  Documentation  📖

**Given** I'm working with an overlay that includes arrays
**When** I read through `ytt` documentation
**Then** it's obvious and clear that the default operation on array items is append

**Given** I'm working with an overlay that includes arrays
**When** I use/explore examples (in the Playground)
**Then** it's obvious and clear that the default operation on array items is append

**Comment by pivotaljohn** [comment]
This issue was copied from a Pivotal Tracker story that incorrectly appeared to be "shovel ready"; it was not.

When ready, we can update _this_ issue and re-open it or create a new one.

**Comment by gcheadle-vmware** [comment]
Code is in 'default-append' branch

**Comment by cari-lynn** [comment]
TODO checklist:
- [x] check if it's possible to use the `#@overlay/match-child-defaults missing_ok=True` annotation on an array item and add a test for it if so.

**Comment by cari-lynn** [comment]
This issue will be available in the v0.32.0 release. When we cut that release, we should merge the documentation PR related to this: https://github.com/vmware-tanzu/carvel.dev/pull/64

**Comment by gcheadle-vmware** [comment]
Release v0.32.0 cut, closing issue.
